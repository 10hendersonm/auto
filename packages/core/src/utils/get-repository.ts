import parseGitHubUrl from "parse-github-url";
import on from "await-to-js";

import execPromise from "./exec-promise";

/**
 * Get the owner and repo from the configure remote "origin"
 */
export default async function getRepository() {
  const [, origin] = await on(
    execPromise("git", ["remote", "get-url", "origin"])
  );

  if (origin) {
    const info =
      parseGitHubUrl(origin) || ({} as Record<string, string | undefined>);
    const { name, owner } = info;

    if (name && owner) {
      return { repo: name, owner };
    }
  }
}
