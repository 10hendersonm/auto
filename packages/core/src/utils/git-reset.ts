import execPromise from "./exec-promise";

/** Reset all git changes */
export const gitReset = async () =>
  execPromise("git", ["reset", "--hard", "HEAD"]);
