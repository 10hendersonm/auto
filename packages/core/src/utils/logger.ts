import signale, { Signale } from 'signale';

export interface ILogger {
  /** The level at which to log messages */
  logLevel: LogLevel;
  /** The default logger. Always on */
  log: signale.Signale<signale.DefaultMethods>;
  /** The verbose log. Has more debug logs */
  verbose: signale.Signale<signale.DefaultMethods>;
  /** The very verbose log. Has all debug logs */
  veryVerbose: signale.Signale<signale.DefaultMethods>;
}

/** Create a dummy logger for testing. */
export function dummyLog(): ILogger {
  return {
    logLevel: undefined,
    log: new Signale({ disabled: true }),
    verbose: new Signale({ disabled: true }),
    veryVerbose: new Signale({ disabled: true })
  };
}

export type LogLevel = 'verbose' | 'veryVerbose' | undefined;

/** Create a logger the the given log level. */
export default function createLog(mode: LogLevel): ILogger {
  return {
    logLevel: mode,
    log: new Signale(),
    verbose: new Signale({
      disabled: mode !== 'verbose' && mode !== 'veryVerbose'
    }),
    veryVerbose: new Signale({
      disabled: mode !== 'veryVerbose'
    })
  };
}
