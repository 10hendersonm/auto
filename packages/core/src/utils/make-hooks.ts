import {
  AsyncParallelHook,
  AsyncSeriesBailHook,
  AsyncSeriesWaterfallHook,
  SyncHook,
  SyncWaterfallHook
} from 'tapable';
import { IAutoHooks } from '../auto';
import { IChangelogHooks } from '../changelog';
import { ILogParseHooks } from '../log-parse';
import { IReleaseHooks } from '../release';

export const makeHooks = (): IAutoHooks => ({
  beforeRun: new SyncHook(['config']),
  modifyConfig: new SyncWaterfallHook(['config']),
  beforeShipIt: new SyncHook([]),
  afterShipIt: new AsyncParallelHook(['version', 'commits']),
  afterRelease: new AsyncParallelHook(['releaseInfo']),
  onCreateRelease: new SyncHook(['options']),
  onCreateChangelog: new SyncHook(['changelog', 'version']),
  onCreateLogParse: new SyncHook(['logParse']),
  getAuthor: new AsyncSeriesBailHook([]),
  getPreviousVersion: new AsyncSeriesBailHook(['prefixRelease']),
  getRepository: new AsyncSeriesBailHook([]),
  version: new AsyncParallelHook(['version']),
  afterVersion: new AsyncParallelHook([]),
  publish: new AsyncParallelHook(['version']),
  afterPublish: new AsyncParallelHook([]),
  canary: new AsyncSeriesBailHook(['canaryVersion', 'postFix'])
});

export const makeReleaseHooks = (): IReleaseHooks => ({
  onCreateChangelog: new SyncHook(['changelog', 'version']),
  createChangelogTitle: new AsyncSeriesBailHook([]),
  onCreateLogParse: new SyncHook(['logParse'])
});

export const makeLogParseHooks = (): ILogParseHooks => ({
  parseCommit: new AsyncSeriesWaterfallHook(['commit']),
  omitCommit: new AsyncSeriesBailHook(['commit'])
});

export const makeChangelogHooks = (): IChangelogHooks => ({
  renderChangelogLine: new AsyncSeriesWaterfallHook(['lineData']),
  renderChangelogTitle: new AsyncSeriesBailHook(['commits', 'lineRender']),
  renderChangelogAuthor: new AsyncSeriesBailHook([
    'author',
    'commit',
    'options'
  ]),
  renderChangelogAuthorLine: new AsyncSeriesBailHook(['author', 'user']),
  omitReleaseNotes: new AsyncSeriesBailHook(['commit'])
});
