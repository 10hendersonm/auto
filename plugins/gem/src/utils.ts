import * as fs from "fs";
import { promisify } from "util";

export const mkdir = promisify(fs.mkdir);
export const readFile = promisify(fs.readFile);
export const writeFile = promisify(fs.writeFile);
