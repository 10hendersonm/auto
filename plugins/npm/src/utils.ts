import * as fs from "fs";
import { promisify } from "util";

export const readFile = promisify(fs.readFile);
export const writeFile = promisify(fs.writeFile);

/** Check if the project is a monorepo */
export const isMonorepo = () => fs.existsSync("lerna.json");

/** Parse the lerna.json file. */
export const getLernaJson = () => {
  try {
    return JSON.parse(fs.readFileSync("lerna.json", "utf8"));
  } catch (error) {
    return {};
  }
};
