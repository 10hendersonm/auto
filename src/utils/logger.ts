import signale, { Signale } from 'signale';

export interface ILogger {
  log: signale.Signale<signale.DefaultMethods>;
  verbose: signale.Signale<signale.DefaultMethods>;
  veryVerbose: signale.Signale<signale.DefaultMethods>;
}

export function dummyLog(): ILogger {
  return {
    log: new Signale({ disabled: true }),
    verbose: new Signale({ disabled: true }),
    veryVerbose: new Signale({ disabled: true })
  };
}

export type LogLevel = 'verbose' | 'veryVerbose' | undefined;

export default function createLog(mode: LogLevel): ILogger {
  return {
    log: new Signale(),
    verbose: new Signale({
      disabled: mode !== 'verbose' && mode !== 'veryVerbose'
    }),
    veryVerbose: new Signale({
      disabled: mode !== 'veryVerbose'
    })
  };
}
