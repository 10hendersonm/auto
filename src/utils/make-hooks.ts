import { AsyncSeriesBailHook, AsyncSeriesHook, SyncHook } from 'tapable';
import { IGitHubReleaseHooks } from '../github-release';
import { IChangelogHooks } from '../log-parse';
import { IAutoHooks } from '../main';

export const makeHooks = (): IAutoHooks => ({
  beforeRun: new SyncHook(['config']),
  beforeShipIt: new SyncHook([]),
  onCreateGitHubRelease: new SyncHook(['config']),
  onCreateChangelog: new SyncHook(['gitHubReleaseConfig']),
  getAuthor: new AsyncSeriesBailHook([]),
  getPreviousVersion: new AsyncSeriesBailHook(['prefixRelease']),
  getRepository: new AsyncSeriesBailHook([]),
  publish: new AsyncSeriesHook(['version'])
});

export const makeGitHubReleaseHooks = (): IGitHubReleaseHooks => ({
  onCreateChangelog: new SyncHook(['changelog'])
});

export const makeChangelogHooks = (): IChangelogHooks => ({
  renderChangelogLine: new AsyncSeriesBailHook(['commits', 'lineRender']),
  renderChangelogTitle: new AsyncSeriesBailHook(['commits', 'lineRender']),
  renderChangelogAuthor: new AsyncSeriesBailHook([
    'author',
    'commit',
    'options'
  ]),
  renderChangelogAuthorLine: new AsyncSeriesBailHook(['author', 'user'])
});
