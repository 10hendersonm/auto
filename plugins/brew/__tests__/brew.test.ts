import Auto from "@auto-it/core";
import { makeHooks } from "@auto-it/core/dist/utils/make-hooks";
import { dummyLog } from "@auto-it/core/dist/utils/logger";
import { execSync } from "child_process";

import Brew from "../src";

const exec = jest.fn();
const existsSync = jest.fn();
const mkdirSync = jest.fn();
const readFileSync = jest.fn();
const writeFileSync = jest.fn();

jest.mock("fs", () => ({
  // @ts-ignore
  read: (a, b, cb) => {
    cb(undefined);
  },
  // @ts-ignore
  existsSync: (...args) => existsSync(...args),
  // @ts-ignore
  mkdirSync: (...args) => mkdirSync(...args),
  // @ts-ignore
  readFileSync: (...args) => readFileSync(...args),
  // @ts-ignore
  writeFileSync: (...args) => writeFileSync(...args),
  // @ts-ignore
  readFile: (a, b, cb) => {
    cb(undefined, "{}");
  },
  // @ts-ignore
  writeFile: (a, b, cb) => {
    cb(undefined, "{}");
  },
  ReadStream: function () {},
  WriteStream: function () {},
  // @ts-ignore
  closeSync: () => undefined,
}));

jest.mock("child_process");

// @ts-ignore
execSync.mockImplementation(exec);

describe("Brew Plugin", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("should do nothing with an uninitialized auto", async () => {
    const brewPlugin = new Brew({ name: "foo", executable: "./foo" });
    const autoHooks = makeHooks();

    brewPlugin.apply({ hooks: autoHooks, logger: dummyLog() } as Auto);

    await autoHooks.afterVersion.promise({});

    expect(exec).not.toHaveBeenCalled();
  });

  test("should exit when there are errors", async () => {
    const exit = jest.fn();
    // @ts-ignore
    process.exit = exit;

    const brewPlugin = new Brew({ name: "foo", executable: "./foo" });
    const autoHooks = makeHooks();

    brewPlugin.apply(({
      hooks: autoHooks,
      logger: dummyLog(),
      git: {
        getLatestTagInBranch: () => {
          throw new Error();
        },
      },
    } as unknown) as Auto);

    await autoHooks.afterVersion.promise({});

    expect(exit).toHaveBeenCalled();
  });

  test("should validate it's configuration", async () => {
    const config = { name: "foo", execuable: "./foo" };
    // @ts-ignore
    const brewPlugin = new Brew(config);
    const autoHooks = makeHooks();

    brewPlugin.apply({ hooks: autoHooks, logger: dummyLog() } as Auto);

    expect(
      await autoHooks.validateConfig.promise("brew", config)
    ).toMatchSnapshot();
  });

  test("should not validate other's configurations", async () => {
    const config = { name: "foo", execuable: "./foo" };
    // @ts-ignore
    const brewPlugin = new Brew(config);
    const autoHooks = makeHooks();

    brewPlugin.apply({ hooks: autoHooks, logger: dummyLog() } as Auto);

    expect(
      await autoHooks.validateConfig.promise("git-tag", config)
    ).toBeUndefined();
  });

  test("should create the formula folder if it does not exist", async () => {
    const brewPlugin = new Brew({ name: "foo", executable: "./foo" });
    const autoHooks = makeHooks();

    brewPlugin.apply(({
      hooks: autoHooks,
      logger: dummyLog(),
      git: { getLatestTagInBranch: () => "v1.0.0" },
    } as unknown) as Auto);

    existsSync.mockReturnValueOnce(false);
    readFileSync.mockReturnValueOnce("");
    exec.mockReturnValueOnce("1234 1234");

    await autoHooks.afterVersion.promise({});

    expect(mkdirSync).toHaveBeenCalled();
  });

  test("should not create the formula folder if it does exist", async () => {
    const brewPlugin = new Brew({ name: "foo", executable: "./foo" });
    const autoHooks = makeHooks();

    brewPlugin.apply(({
      hooks: autoHooks,
      logger: dummyLog(),
      git: { getLatestTagInBranch: () => "v1.0.0" },
    } as unknown) as Auto);

    existsSync.mockReturnValueOnce(true);
    readFileSync.mockReturnValueOnce("");
    exec.mockReturnValueOnce("1234 1234");

    await autoHooks.afterVersion.promise({});

    expect(mkdirSync).not.toHaveBeenCalled();
  });

  test("should create formula", async () => {
    const brewPlugin = new Brew({ name: "foo", executable: "./foo" });
    const autoHooks = makeHooks();

    brewPlugin.apply(({
      hooks: autoHooks,
      logger: dummyLog(),
      git: { getLatestTagInBranch: () => "v1.0.0" },
    } as unknown) as Auto);

    readFileSync.mockReturnValueOnce("$VERSION $SHA");
    exec.mockReturnValueOnce("1234 1234");

    await autoHooks.afterVersion.promise({});

    expect(writeFileSync).toHaveBeenCalledWith(
      "./Formula/foo.rb",
      "v1.0.0 1234"
    );
  });

  test("should create multiple formulae", async () => {
    const brewPlugin = new Brew([
      { name: "foo", executable: "./foo" },
      { name: "bar", executable: "./bar" },
    ]);
    const autoHooks = makeHooks();

    brewPlugin.apply(({
      hooks: autoHooks,
      logger: dummyLog(),
      git: { getLatestTagInBranch: () => "v1.0.0" },
    } as unknown) as Auto);

    readFileSync.mockReturnValue("$VERSION $SHA");
    exec.mockReturnValue("1234 1234");

    await autoHooks.afterVersion.promise({});

    expect(writeFileSync).toHaveBeenCalledWith(
      "./Formula/foo.rb",
      "v1.0.0 1234"
    );
    expect(writeFileSync).toHaveBeenCalledWith(
      "./Formula/bar.rb",
      "v1.0.0 1234"
    );
  });
});
