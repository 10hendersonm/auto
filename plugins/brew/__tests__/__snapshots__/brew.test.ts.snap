// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Brew Plugin should validate it's configuration 1`] = `
Array [
  Object {
    "expectedType": "\\"string\\"",
    "path": "brew.executable",
    "value": undefined,
  },
]
`;
